### PR TITLE
nixos/hddtemp: add support for HDD/SSD temperature montoring

### PR DESCRIPTION
**Pull request by peterhoeg**
###### Motivation for this change

Basic module for temperature monitoring + some minor `hddtemp` cleanups while at it.

The module does a few things:
1. enables you to add drive database entries
2. resolves drive symlinks which is needed if you want to pass a drive `by-path` as hddtemp chokes on colons
3. because we run this from a wrapper script, we can pass wildcards for the drives which is otherwise not possible directly using `ExecStart`.
3. limits the permissions somewhat

Ideally we lock this down properly with `DynamicUser = true;` but the problem is permissions needed to fetch SMART data which doesn't seem doable with the existing capabilities.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by bjornfor — APPROVED** [review]

File: pkgs/tools/misc/hddtemp/default.nix [hunk 1]
```diff
@@ -1,13 +1,15 @@
...
     sha256 = "1fr6qgns6qv7cr40lic5yqwkkc7yjmmgx8j0z6d93csg3smzhhya";
   };
 
+in
+stdenv.mkDerivation rec {
+  pname = "hddtemp";
+  version = "0.3_beta15";
+
   src = fetchurl {
     url = "mirror://savannah/hddtemp/hddtemp-0.3-beta15.tar.bz2";
```
Comment (sternenseemann) [inline pkgs/tools/misc/hddtemp/default.nix:14]: Minor thing: You could also tie this to `version`.

```suggestion
    url = "mirror://savannah/hddtemp/hddtemp-${builtins.replaceStrings [ "_" ] [ "-" ] version}.tar.bz2";
```

File: nixos/modules/hardware/sensor/hddtemp.nix [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    set -eEuo pipefail
+
+    file=/var/lib/hddtemp/hddtemp.db
+
+    drives=(${toString (map (e: ''$(realpath "${e}") '') cfg.drives)})
+
+    cp ${pkgs.hddtemp}/share/hddtemp/hddtemp.db $file
+    ${lib.concatMapStringsSep "\n" (e: "echo '${e}' >> $file") cfg.dbEntries}
+
+    exec ${pkgs.hddtemp}/bin/hddtemp ${lib.concatStringsSep " " cfg.extraArgs} \
```
Comment (sternenseemann) [inline nixos/modules/hardware/sensor/hddtemp.nix:17]: Consider using `lib.escapeShellArgs` which would prevent users from accidentally messing up the command line. Idk if there may be a genuine use for variables or something here though.

```suggestion
    exec ${pkgs.hddtemp}/bin/hddtemp ${lib.escapeShellArgs cfg.extraArgs} \
```

File: nixos/modules/hardware/sensor/hddtemp.nix [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+in
+{
+  meta.maintainers = with lib.maintainers; [ peterhoeg ];
+
+  ###### interface
+
+  options = {
+    hardware.sensor.hddtemp = {
+      enable = mkOption {
```
Comment (sternenseemann) [inline nixos/modules/hardware/sensor/hddtemp.nix:32]: You could also use `mkEnableOption` here.

File: pkgs/tools/misc/hddtemp/default.nix (hunk 1 above)
Comment (peterhoeg) [reply pkgs/tools/misc/hddtemp/default.nix]: Good catch. There's actually no reason why the version should have an underscore to begin with, so I fixed that up instead.

**Comment by peterhoeg** [comment]
@GrahamcOfBorg  build hddtemp

File: nixos/modules/hardware/sensor/hddtemp.nix [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  inherit (lib) mkIf mkOption types;
+
+  cfg = config.hardware.sensor.hddtemp;
+
+  wrapper = pkgs.writeShellScript "hddtemp-wrapper" ''
+    set -eEuo pipefail
+
+    file=/var/lib/hddtemp/hddtemp.db
+
+    drives=(${toString (map (e: ''$(realpath "${e}") '') cfg.drives)})
```
Comment (sternenseemann) [inline nixos/modules/hardware/sensor/hddtemp.nix:12]:
```suggestion
    drives=(${toString (map (e: ''$(realpath ${lib.escapeShellArg e}) '') cfg.drives)})
```

File: nixos/modules/hardware/sensor/hddtemp.nix [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+    cp ${pkgs.hddtemp}/share/hddtemp/hddtemp.db $file
+    ${lib.concatMapStringsSep "\n" (e: "echo '${e}' >> $file") cfg.dbEntries}
```
Comment (sternenseemann) [inline nixos/modules/hardware/sensor/hddtemp.nix:15]:
```suggestion
    ${lib.concatMapStringsSep "\n" (e: "echo ${lib.escapeShellArg e} >> $file") cfg.dbEntries}
```

**Review by sternenseemann — CHANGES_REQUESTED** [review]
LGTM apart from a few escaping improvements.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 111383` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package blacklisted:
</summary>
<ul>
<li>tests.nixos-functions.nixos-test</li>
</ul>
</details>
<details>
<summary>
3 packages built:
</summary>
<ul>
<li>glances</li>
<li>hddtemp</li>
<li>xfce.xfce4-sensors-plugin (xfce.xfce4_sensors_plugin)</li>
</ul>
</details>

File: nixos/modules/hardware/sensor/hddtemp.nix (hunk 2 above)
Comment (peterhoeg) [reply nixos/modules/hardware/sensor/hddtemp.nix]: Good point regd escaping - all sorted

**Review by sternenseemann — APPROVED** [review]
LGTM

**Pull request merged**